src/Infrastructure.EntityFramework/EntityFrameworkCache.cs

```diff
@@ -0,0 +1,304 @@
+    {
+        var utcNow = DateTime.UtcNow;
+
+        // resolve options
+        if (!options.AbsoluteExpiration.HasValue &&
+            !options.AbsoluteExpirationRelativeToNow.HasValue &&
+            !options.SlidingExpiration.HasValue)
+        {
+            options = new DistributedCacheEntryOptions
+            {
+                SlidingExpiration = _defaultSlidingExpiration
+            };
+        }
+
+        if (cache == null)
+        {
+            // do an insert
+            cache = new Cache { Id = key };
+        }
+
+        var slidingExpiration = (long?)options.SlidingExpiration?.TotalSeconds;
+
+        // calculate absolute expiration
+        DateTime? absoluteExpiration = null;
+        if (options.AbsoluteExpirationRelativeToNow.HasValue)
+        {
+            absoluteExpiration = utcNow.Add(options.AbsoluteExpirationRelativeToNow.Value);
+        }
+        else if (options.AbsoluteExpiration.HasValue)
+        {
+            if (options.AbsoluteExpiration.Value <= utcNow)
+            {
+                throw new InvalidOperationException("The absolute expiration value must be in the future.");
+            }
+
+            absoluteExpiration = options.AbsoluteExpiration.Value.DateTime;
+        }
+
+        // set values on cache
+        cache.Value = value;
+        cache.SlidingExpirationInSeconds = slidingExpiration;
+        cache.AbsoluteExpiration = absoluteExpiration;
+        if (slidingExpiration.HasValue)
+        {
+            cache.ExpiresAtTime = utcNow.AddSeconds(slidingExpiration.Value);
+        }
+        else if (absoluteExpiration.HasValue)
+        {
+            cache.ExpiresAtTime = absoluteExpiration.Value;
+        }
+        else
+        {
+            throw new InvalidOperationException("Either absolute or sliding expiration needs to be provided.");
+        }
+
+        return cache;
+    }
+
+    private bool UpdateCacheExpiration(Cache cache)
+    {
+        var utcNow = DateTime.UtcNow;
+        if (cache.SlidingExpirationInSeconds.HasValue && (cache.AbsoluteExpiration.HasValue || cache.AbsoluteExpiration != cache.ExpiresAtTime))
+        {
+            if (cache.AbsoluteExpiration.HasValue && (cache.AbsoluteExpiration.Value - utcNow).TotalSeconds <= cache.SlidingExpirationInSeconds)
+            {
+                cache.ExpiresAtTime = cache.AbsoluteExpiration.Value;
+            }
+            else
+            {
+                cache.ExpiresAtTime = utcNow.AddSeconds(cache.SlidingExpirationInSeconds.Value);
+            }
+            return true;
+        }
+        return false;
+    }
+
+    private void ScanForExpiredItemsIfRequired()
+    {
+        lock (_mutex)
+        {
+            var utcNow = DateTime.UtcNow;
+            if ((utcNow - _lastExpirationScan) > _expiredItemsDeletionInterval)
+            {
+                _lastExpirationScan = utcNow;
+                Task.Run(_deleteExpiredCachedItemsDelegate);
```

Comment:
This concerns me a bit, I know this is how the SqlServer one does it but a DbContext is only supposed to be accessed from a single thread at a time and it has safety mechanisms to detect that. Read the bullet point three in [this section](https://learn.microsoft.com/en-us/ef/core/performance/advanced-performance-topics?tabs=with-di%2Cexpression-api-with-constant#reducing-runtime-overhead). We might be okay because it's a different instance from a different scope but I'd really love some tests making sure this works.

Reply:
There's a whole bunch of tests from MS on this here: https://github.com/dotnet/aspnetcore/tree/main/src/Caching/SqlServer/test

Do you see a way we can bring these over for our implementation?